### PR TITLE
Download CSV for Experience Points of all Students in Course

### PR DESCRIPTION
**Pull request by bivanalhar**
Some notable changes within this PR:
- Modifying the sidebar Disburse Points to become Experience Points due to the additional feature of overseeing the whole experience points in the whole course made available in one table
- Inside the page, add one more tab called Experience Points Details
- Having two more options of Download CSV and also Filter by Name, in case user wants to see the experience points history of one particular users without needing to navigate to other page
- Since we already have the page containing experience points history of each user, we allow user to navigate to that page by simply clicking the name inside each row
- There are differences in ordering for the display and also for the CSV. In the display, we order it based on the Updated Time, so that if user have just updated the experience points, they can check whether the points have been updated quite easily
- Meanwhile, for CSV, the ordering is based on the name of the students so that it's easier for user to look into the points of each students
- Also, if user applies for some filtering, then the CSV downloaded will only contain the data of that identified student's experience points

Screenshot 2023-10-03 at 5 25 23 PM

- Applying the filtering will immediately render the table with only the specified name's entries, as shown in the example below
Screenshot 2023-10-03 at 5 32 22 PM

- Clearing the filter will immediately remove the mentioned filtering and reset the table to show all the entries indicating all experience points achieved within that course for all users

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -205,3 +205,5 @@ gem 'rails-html-sanitizer', '>= 1.0.4'
 
 gem 'mimemagic', '0.4.3'
 gem 'ffi', '>= 1.14.2'
+
+gem 'rubocop', '~> 1.56'
```
Comment (ekowidianto) [inline Gemfile:209]: Why do we need to add this now?

File: client/app/bundles/course/experience-points/store.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  downloadExperiencePointsFailure:
+    (): DownloadExperiencePointsFailureAction => {
+      return {
+        type: DOWNLOAD_EXPERIENCE_POINTS_FAILURE,
+      };
+    },
+};
+
+export default reducer;
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/store.ts:141]: use redux toolkit. See other codebase with `createSlice`

**Review by ekowidianto — COMMENTED** [review]
@bivanalhar Some initial comments

File: app/models/components/course/course_ability_component.rb [hunk 3]
```diff
@@ -37,6 +38,11 @@ def allow_staff_show_course_users
     can :show_users, Course, id: course.id
   end
 
+  def allow_staff_read_all_experience_points
```
Comment (ekowidianto) [inline app/models/components/course/course_ability_component.rb:41]: You try to masquerade as a student, run this in `http://lvh.me:8080/courses/{course_number}/experience_points_records?format=json` and see if your backend authorization works properly

File: app/models/components/course/experience_points_records_ability_component.rb [hunk 4]
```diff
@@ -22,5 +22,6 @@ def allow_read_course_experience_points_records
 
   def allow_read_own_experience_points_records
     can :read, Course::ExperiencePointsRecord, course_user: { user_id: user.id }
+    can :show_user_exp, Course::ExperiencePointsRecord, course_user: { user_id: user.id }
```
Comment (ekowidianto) [inline app/models/components/course/experience_points_records_ability_component.rb:25]: Do we need both this and the one above?

File: app/controllers/course/experience_points_records_controller.rb [hunk 5]
```diff
@@ -1,21 +1,42 @@
...
 class Course::ExperiencePointsRecordsController < Course::ComponentController
-  load_resource :course_user, through: :course, id_param: :user_id
+  load_resource :course_user, through: :course, id_param: :user_id, except: [:read_all_exp, :download]
   load_and_authorize_resource :experience_points_record, through: :course_user,
-                                                         class: Course::ExperiencePointsRecord.name
+                                                         class: Course::ExperiencePointsRecord.name,
+                                                         except: [:read_all_exp, :download]
+
+  def read_all_exp
+    if filter_and_page_params[:student_id].present?
```
Comment (ekowidianto) [inline app/controllers/course/experience_points_records_controller.rb:9]: To make it more succinct, why not
`course_user_id = filter_and_page_params[:student_id].present? ? filter_and_page_params[:student_id] : @course.course_users.pluck(:id)`
`@experience_points_records = Course::ExperiencePointsRecord.where(course_user_id: course_user_id)`

Comment (ekowidianto) [reply app/controllers/course/experience_points_records_controller.rb:9]: And why duplicate code with `fetch_experience_points_records`?

File: client/app/api/course/ExperiencePointsRecord.ts [hunk 6]
```diff
@@ -11,45 +13,57 @@ export default class ExperiencePointsRecordAPI extends BaseCourseAPI {
...
     return this.client.get(
-      `${
-        this.#urlPrefix
-      }/users/${userId}/experience_points_records?filter[page_num]=${pageNum}`,
+      `${this.#urlPrefix}/users/${userId}/experience_points_records`,
+      { params: { 'filter[page_num]': pageNum } },
     );
   }
 
+  download(studentId?: number): Promise<AxiosResponse<JobSubmitted>> {
```
Comment (ekowidianto) [inline client/app/api/course/ExperiencePointsRecord.ts:41]: `downloadCSV`

File: client/app/api/course/ExperiencePointsRecord.ts [hunk 7]
```diff
@@ -11,45 +13,57 @@ export default class ExperiencePointsRecordAPI extends BaseCourseAPI {
     return `/courses/${this.courseId}`;
   }
 
+  /**
+   * Fetches all experience points records from all user
+   */
+  indexAll(
```
Comment (ekowidianto) [inline client/app/api/course/ExperiencePointsRecord.ts:19]: use a more consistent naming similar to backend. same for `index` below

File: client/app/api/course/ExperiencePointsRecord.ts [hunk 8]
```diff
@@ -11,45 +13,57 @@ export default class ExperiencePointsRecordAPI extends BaseCourseAPI {
...
+    studentId?: number,
+    pageNum: number = 1,
+  ): Promise<AxiosResponse<ExperiencePointsRecords>> {
```
Comment (ekowidianto) [inline client/app/api/course/ExperiencePointsRecord.ts:22]: use APIResponse

File: client/app/api/course/ExperiencePointsRecord.ts [hunk 9]
```diff
@@ -11,45 +13,57 @@ export default class ExperiencePointsRecordAPI extends BaseCourseAPI {
...
    * Update an experience points record for a user
    */
   update(
     params: UpdateExperiencePointsRecordPatchData,
     recordId: number,
+    studentId?: number,
   ): Promise<AxiosResponse<ExperiencePointsRecordListData>> {
     const url = `${this.#urlPrefix}/users/${
-      this.courseUserId
+      this.courseUserId ?? studentId
```
Comment (ekowidianto) [inline client/app/api/course/ExperiencePointsRecord.ts:56]: I suggest to not use `this.courseUserId` at all so `studentId` is not optional anymore. In this case, we dont try to infer the api path from existing url. Same for delete below.

So, `updateExperiencePointsRecord` and `deleteExperiencePointsRecord` also require `studentId`.
Instead of `studentId`, call it `courseUserId` to be more generic.

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx [hunk 10]
```diff
@@ -5,22 +5,22 @@ import {
...
-  updateExperiencePointsRecord,
-} from '../../operations';
-
 interface Props extends WrappedComponentProps {
   permissions: ExperiencePointsRecordPermissions;
   data: ExperiencePointsRowData;
   isDirty: boolean;
   isManuallyAwarded: boolean;
   handleSave: (newData: ExperiencePointsRowData) => void;
+  studentId?: number;
```
Comment (ekowidianto) [inline client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:23]: Make this non optional?

File: app/views/course/experience_points_records/read_all_exp.json.jbuilder [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+json.rowCount @experience_points_count
+json.records @experience_points_records.includes(:course_user) do |record|
+  json.partial! 'experience_points_record', course: current_course, record: record
+  json.courseUserName record.course_user.name
```
Comment (ekowidianto) [inline app/views/course/experience_points_records/read_all_exp.json.jbuilder:5]: Dont you have `json.student` inside `experience_points_record` partial already? Same for `json.userExperienceUrl` below

File: app/views/course/experience_points_records/read_all_exp.json.jbuilder [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  json.userExperienceUrl course_user_experience_points_records_path(current_course, record.course_user_id)
+end
+
+students = current_course.course_users.order_alphabetically.student
+json.filters do
+  json.names students do |student|
```
Comment (ekowidianto) [inline app/views/course/experience_points_records/read_all_exp.json.jbuilder:11]: names is not specific enough. What type of names are you referring here? I would use `courseUser` instead.

File: app/jobs/course/experience_points_download_job.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+class Course::ExperiencePointsDownloadJob < ApplicationJob
+  include TrackableJob
+  queue_as :lowest
+
+  protected
+
+  def perform_tracked(current_course, course_user_id)
+    csv_file = Course::ExperiencePointsDownloadService.
+               download(current_course, course_user_id)
+    redirect_to SendFile.send_file(csv_file)
```
Comment (ekowidianto) [inline app/jobs/course/experience_points_download_job.rb:11]: Add a proper `public_name`? ie `{course_name}_experience_records.csv`

File: app/services/course/experience_points_download_service.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            I18n.t('course.experience_points_records.download.name'),
+            I18n.t('course.experience_points_records.download.updater'),
+            I18n.t('course.experience_points_records.download.reason'),
+            I18n.t('course.experience_points_records.download.exp_points')]
+  end
+
+  def download_exp_points(csv, record)
+    point_updater = @updater_preload_service.course_user_for(record.updater) || record.updater
+
+    @reason = if record.manually_awarded?
```
Comment (ekowidianto) [inline app/services/course/experience_points_download_service.rb:54]: Why use instance variable here instead of just normal variable?

File: app/services/course/experience_points_download_service.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                case record.specific.actable
+                when Course::Assessment::Submission
+                  record.specific.assessment.title
+                when Course::Survey::Response
+                  record.specific.survey.title
+                end
+              end
+
+    csv << [record.updated_at,
+            record.course_user.name,
```
Comment (ekowidianto) [inline app/services/course/experience_points_download_service.rb:66]: Wouldnt this trigger N+1?

File: Gemfile (hunk 1 above)
Comment (bivanalhar) [reply Gemfile:209]: it's quite annoying if I want to use rubocop in my local, I need to do `bundle add rubocop` all the time, and then in the push, I need to remove that..

I think it's also useful to be used before creating the PR, so that hound will give feedback much lesser

Comment (ekowidianto) [reply Gemfile:209]: I dont need to use bundle add rubocop and still can run rubocop?

Comment (ekowidianto) [reply Gemfile:209]: Alternatively, do we need rubocop in production? Why not put it under development group?

File: app/models/components/course/experience_points_records_ability_component.rb (hunk 4 above)
Comment (bivanalhar) [reply app/models/components/course/experience_points_records_ability_component.rb:25]: at first, I tried deleting that `can :read` line and as a result, even student cannot see the Exp box in the UserProfileCard..

Now I look again, I immediately know why is that so, and thus that line's not needed anymore

File: Gemfile (hunk 1 above)
Comment (bivanalhar) [reply Gemfile:209]: not really sure, maybe need some more configuration on my local machine for this matter

in the meanwhile, I put the rubocop under development group, as suggested

File: client/app/api/course/ExperiencePointsRecord.ts (hunk 9 above)
Comment (bivanalhar) [reply client/app/api/course/ExperiencePointsRecord.ts]: I'm not quite sure actually that courseUserId is more intuitive than studentId. Indeed, the url formed has prefixes of users, but in here, the owner of the exp points are student itself

The reason I need to put emphasis on student is because inside the entity of exp points, there is another object similar with student as well (updater) and thus I need to differentiate if the id mentioned belongs to student or updater

File: app/views/course/experience_points_records/read_all_exp.json.jbuilder (hunk 11 above)
Comment (bivanalhar) [reply app/views/course/experience_points_records/read_all_exp.json.jbuilder]: apparently these 2 fields are not used anymore (I have changed completely the usage of record.courseUserName into record.student.name, same with Url).

File: client/app/routers/AuthenticatedApp.tsx [hunk 16]
```diff
@@ -244,6 +244,11 @@ const authenticatedRouter: Translated<RouteObject[]> = (t) =>
             },
           ],
         },
+        {
```
Comment (ekowidianto) [inline client/app/routers/AuthenticatedApp.tsx:247]: Any reason you moved the sequence?

Comment (bivanalhar) [reply client/app/routers/AuthenticatedApp.tsx:247]: With the current structure inside redux being modified (now experience points is not anymore coupled with users), and the Experience Points page being not only containing disbursement and forum but also the info regarding **all users**’ exp points, it’s only natural that the routes structure needs to be changed also (to get it out from users and only being a sub of course)

File: app/controllers/course/experience_points_records_controller.rb [hunk 17]
```diff
@@ -1,21 +1,38 @@
...
 class Course::ExperiencePointsRecordsController < Course::ComponentController
-  load_resource :course_user, through: :course, id_param: :user_id
+  load_resource :course_user, through: :course, id_param: :user_id, except: [:read_all_exp, :download]
   load_and_authorize_resource :experience_points_record, through: :course_user,
-                                                         class: Course::ExperiencePointsRecord.name
+                                                         class: Course::ExperiencePointsRecord.name,
+                                                         except: [:read_all_exp, :download]
+
+  def read_all_exp
+    authorize!(:read_all_exp, @course)
```
Comment (ekowidianto) [inline app/controllers/course/experience_points_records_controller.rb:9]: Why are you authorising to read the course?

Comment (ekowidianto) [reply app/controllers/course/experience_points_records_controller.rb:9]: Nevermind i will refactor

Comment (bivanalhar) [reply app/controllers/course/experience_points_records_controller.rb:9]: experiences now belong to the course, and hence the ability whether user can read that page (containing all the experience points) lies in the course-level..

File: spec/features/course/experience_points/forum_disbursement_spec.rb [hunk 18]
```diff
@@ -28,7 +28,8 @@
                                           created_at: 3.weeks.ago, updated_at: 3.weeks.ago)
         end
         create(:course_discussion_post_vote, post: older_posts[0])
-        visit disburse_experience_points_course_users_path(course)
+        visit course_experience_points_records_path(course)
+        find('button#forum-disbursement-tab').click
```
Comment (ekowidianto) [inline spec/features/course/experience_points/forum_disbursement_spec.rb:32]: Find by the text instead of the class name

File: config/locales/en/course/experience_points_records.yml [hunk 19]
```diff
@@ -2,4 +2,11 @@ en:
   course:
     experience_points_records:
       index:
-        header: 'Experience Points History for %{name}'
+        header: "Experience Points History for %{name}"
```
Comment (ekowidianto) [inline config/locales/en/course/experience_points_records.yml:5]: Revert to single quote

File: client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    id: 'course.experiencePoints.filterByNameButton',
+    defaultMessage: 'Filter by Name',
+  },
+  downloadCsvButton: {
+    id: 'course.experiencePoints.downloadCsvButton',
+    defaultMessage: 'Download CSV',
+  },
+});
+
+const ExperiencePointsFilterDownload: FC<Props> = (props) => {
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx:43]: Disable when downloading/no record.

File: client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+          <IconButton disabled={disabled} onClick={onClick} size="small">
+            <Download />
+          </IconButton>
+        </Tooltip>
+      </Grid>
+    </div>
+  );
+};
+
+export default ExperiencePointsFilterDownload;
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx:101]: Separate this into 2 components. Filter and Download. You need to separate their responsibilities. remember SRP?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTable.tsx [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+  const { t } = useTranslation();
+
+  const records = useAppSelector(getAllExpPointsRecordsEntities);
+
+  const fetchExperiencePoints = isStudentPage
+    ? fetchUserExperiencePointsRecord(studentId!, pageNum)
+    : fetchAllExperiencePointsRecord(studentId, pageNum);
+
+  useEffect(() => {
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTable.tsx:56]: The responsibility of `ExperiencePointsTable` is merely to display the data. The data fetching should not be mixed. There needs to be a separation of responsibility between components here.

Now as you have seen, because of that, you need to pass isStudentPage and check for studentId. You have if else and you are forcing studentId to be non-undefined in line 53 above.

I know it is based on the old-code but in this case, the old code is not correct.

File: client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+  } = props;
+
+  const { t } = useTranslation();
+
+  return (
+    <div className="flex w-full justify-between">
+      <Grid columns={{ xs: 1, md: 3 }} container>
+        <Grid item paddingBottom={1} paddingRight={1} xs={1}>
+          <Autocomplete
+            key="experience-points-name-selector"
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx:61]: Do we need the key here?

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx [hunk 24]
```diff
@@ -5,22 +5,24 @@ import {
...
-
 interface Props extends WrappedComponentProps {
   permissions: ExperiencePointsRecordPermissions;
   data: ExperiencePointsRowData;
   isDirty: boolean;
   isManuallyAwarded: boolean;
   handleSave: (newData: ExperiencePointsRowData) => void;
+  studentId: number;
+  isErrorInput: boolean;
+  isDownloading?: boolean;
```
Comment (ekowidianto) [inline client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:25]: PointManagementButtons should not know the existence of download or whether there's an error. Instead, you can pass `disabled` as a prop and add the logic when calling this component. So you have something like `disabled = isErrorInput || isDownloading || isDirty`.

Better to use disabled as props and it should be non-optional.

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx [hunk 25]
```diff
@@ -48,20 +50,29 @@ const translations = defineMessages({
...
   const dispatch = useAppDispatch();
   const [isSaving, setIsSaving] = useState(false);
   const [isDeleting, setIsDeleting] = useState(false);
 
   const onSave = (): void => {
     setIsSaving(true);
-    dispatch(updateExperiencePointsRecord(data))
-      .then((response) => {
+    dispatch(updateExperiencePointsRecord(data, studentId))
+      .then(() => {
```
Comment (ekowidianto) [inline client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:71]: Why is this removed?

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx [hunk 26]
```diff
@@ -103,7 +114,15 @@ const PointManagementButtons: FC<Props> = (props) => {
...
         <SaveButton
           className={`record-save-${data.id}`}
-          disabled={isSaving || isDeleting || !isDirty}
+          disabled={
+            isSaving ||
+            isDeleting ||
+            !isDirty ||
+            isErrorInput ||
+            Number.isNaN(Number(data.pointsAwarded)) ||
+            !data.pointsAwarded ||
```
Comment (ekowidianto) [inline client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:123]: What's the difference between this and the one above?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 27]
```diff
@@ -0,0 +1,196 @@
...
+    id,
+    reason: '',
+    pointsAwarded: 0,
+  } as ExperiencePointsRowData);
+  const [defaultRowData, setDefaultRowData] = useState({
+    id,
+    reason: '',
+    pointsAwarded: 0,
+  } as ExperiencePointsRowData);
+  const isNotManuallyAwarded = maxExp;
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:38]: Why is this called `isNotManuallyAwarded` but it takes the value of `maxExp`?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 28]
```diff
@@ -0,0 +1,196 @@
...
+import NumberTextField from 'lib/components/core/fields/NumberTextField';
+import TextField from 'lib/components/core/fields/TextField';
+import Link from 'lib/components/core/Link';
+import { formatMiniDateTime } from 'lib/moment';
+
+interface Props {
+  id: number;
+  isStudentPage?: boolean;
+  record: ExperiencePointsRecordMiniEntity;
+  maxExp?: number;
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:19]: Why are you passing this? You can get it from the record directly no?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 29]
```diff
@@ -0,0 +1,196 @@
...
+} from 'types/course/experiencePointsRecords';
+
+import PointManagementButtons from 'course/users/components/buttons/PointManagementButtons';
+import NumberTextField from 'lib/components/core/fields/NumberTextField';
+import TextField from 'lib/components/core/fields/TextField';
+import Link from 'lib/components/core/Link';
+import { formatMiniDateTime } from 'lib/moment';
+
+interface Props {
+  id: number;
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:16]: Now i realise, i think this one can be retrieved from record as well. Can help to remove this?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+          isManuallyAwarded={record.reason.isManuallyAwarded}
+          permissions={record.permissions}
+          studentId={record.student.id}
+        />
+      </TableCell>
+    </TableRow>
+  );
+};
+
+export default memo(
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:191]: memo(ExperiencePointsTableRow, equal) works no?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+    }
+    return rowData.reason;
+  };
+
+  return (
+    <TableRow key={record.id} hover id={`record-${record.id}`}>
+      <TableCell>{formatMiniDateTime(record.updatedAt)}</TableCell>
+      {!isStudentPage && (
+        <TableCell>
+          {record.student.userUrl ? (
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:125]: Follow line 134 below?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx [hunk 32]
```diff
@@ -0,0 +1,196 @@
...
+      const updateData: ExperiencePointsRowData = {
+        ...newData,
+        pointsAwarded: Number(newData.pointsAwarded),
+      };
+      setDefaultRowData({ ...updateData });
+      setIsDirty(false);
+    }
+  };
+
+  const renderReason = (): JSX.Element | string => {
```
Comment (ekowidianto) [inline client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:87]: Create its own component for this: `ReasonCell` or something.

Same for your `<NumberTextField />` for the points cell.

That way, you dont have too many different usestate and logic inside this one component.

**Review by ekowidianto — CHANGES_REQUESTED** [review]
@bivanalhar Please find some comments below. I have pushed some refactoring. Please do the following:
- Address the comments
- Help me fix the tests if anything fails
- Squash the commits (We dont want to have 60 commits for this PR)

I will then do a final review then.

Thanks!

File: client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx (hunk 23 above)
Comment (bivanalhar) [reply client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx:61]: hmm this is weird. last time I removed the key, the linter for this file produces an error that key is needed. Now seems like not anymore

Comment (cysjonathan) [reply client/app/bundles/course/experience-points/ExperiencePointsFilterDownload.tsx:61]: A “key” is a special string attribute you need to include when creating lists of elements in React.
There's no list here, so shouldn't be needed.
Then if its not needed then why don't we remove it before resolving?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTable.tsx (hunk 22 above)
Comment (bivanalhar) [reply client/app/bundles/course/experience-points/components/ExperiencePointsTable.tsx:56]: I have moved this component accordingly
however, this indicator of whether the component is called in student page or not is still important, since the table displayed at `course/{course_id}/experience_points_records` and `course/{course_id}/users/{user_id}/experience_points_records` has one subtle differences (in the latter, we don't need to display the student's name in each row, but we need to do that in the former one)

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx (hunk 25 above)
Comment (bivanalhar) [reply client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:71]: because it's not needed anymore in the subsequent operation..

File: client/app/types/course/experiencePointsRecords.ts [hunk 33]
```diff
@@ -50,7 +60,7 @@ export interface ExperiencePointsRecordMiniEntity {
 export interface ExperiencePointsRowData {
   id: number;
   reason: string;
-  pointsAwarded: number;
+  pointsAwarded: number | string;
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/types/course/experiencePointsRecords.ts:62]: yes.. this is also used to store the current value typed by the user in the rowTable (before saving), and user might type either "-" (intent is for negative number) or "." (intent is for decimal), which is definitely not a number

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx (hunk 27 above)
Comment (bivanalhar) [reply client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:38]: the intention is basically to store the boolean value in which maxExp is defined or not (if it has value, then that isNotManuallyAwarded is true, otherwise false)..

for clarity, I changed the logic a bit to emphasise on that

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx (hunk 25 above)
Comment (ekowidianto) [reply client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:71]: Why do we not use the backend response data in updating the state?

File: client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx (hunk 27 above)
Comment (ekowidianto) [reply client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx:38]: But there’s ismanuallyawarded in pointsreason?

File: client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx (hunk 25 above)
Comment (bivanalhar) [reply client/app/bundles/course/users/components/buttons/PointManagementButtons.tsx:71]: this can be done. I have refactored properly

File: app/models/course/experience_points_record.rb [hunk 34]
```diff
@@ -74,4 +75,45 @@ def set_awarded_attributes
     self.awarded_at ||= Time.zone.now
     self.awarder ||= User.stamper
   end
+
+  def validate_limit_exp_points_on_association
+    return if manually_awarded?
+
+    case specific.actable
+    when Course::Assessment::Submission
+      return unless specific.assessment
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/models/course/experience_points_record.rb:84]: previously it has error, I thought it's because assessment's not tied to submission but apparently it's because the points awarded is nil. After doing the fix, I forgot to delete this. Really sorry, now it's deleted already

**Review by ekowidianto — APPROVED** [review]
@bivanalhar Good job - much better!

Take a look at the changes i made thanks
